### ISSUE TITLE
In "Session" mode client_encoding is set twice?

### ISSUE THREAD

**Issue by pyrocks** [pgbouncer/pgbouncer#406]
Hello,

A little background: I am testing PgBouncer as a connection pooler for AWS Lambda (short lived containers) connecting to Postgres on AWS RDS (Amazon's managed service).
Since each lambda invocation is async and independent of other lambdas - it creates its own connection and closes it when it is finished.

I tried benchmarking pgbouncer 1.9 (EDIT: also in 1.10) in both `Session` and `Transaction` and I noticed that in `Session` there are 2 calls to set client_encoding vs only 1 in `Transaction`. Why does it bother me? Because I'm purposely benchmarking it on a DB in a remote region for which every query takes ~200ms.
The benchmark code is a simple: connect --> select 1 --> select 2 --> disconnect

Log with `Transaction`:
```
2019-08-12 06:44:01.007 1 DEBUG C-0x56245456b010: (nodb)/(nouser)@x.x.x.x:54256 got var: user=someuser
2019-08-12 06:44:01.007 1 DEBUG C-0x56245456b010: (nodb)/(nouser)@x.x.x.x:54256 got var: database=somedb
2019-08-12 06:44:01.007 1 DEBUG C-0x56245456b010: (nodb)/(nouser)@x.x.x.x:54256 got var: client_encoding='utf-8'
2019-08-12 06:44:01.007 1 LOG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 login attempt: db=somedb user=someuser tls=no
2019-08-12 06:44:01.007 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 P: got connection: x.x.x.x:54256 -> x.x.x.x:5432
2019-08-12 06:44:01.008 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 logged in
2019-08-12 06:44:01.009 1 DEBUG S-0x5624545bc350: somedb/someuser@y.y.y.y:5432 varcache_apply: SET client_encoding='''utf-8''';
2019-08-12 06:44:01.212 1 DEBUG S-0x5624545bc350: somedb/someuser@y.y.y.y:5432 S: param: client_encoding = UTF8
2019-08-12 06:44:01.212 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 setting client var: client_encoding='UTF8'
2019-08-12 06:44:01.415 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 query time: 405708 us
2019-08-12 06:44:01.415 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 transaction time: 405708 us
2019-08-12 06:44:01.620 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 query time: 202784 us
2019-08-12 06:44:01.620 1 DEBUG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 transaction time: 202784 us
2019-08-12 06:44:01.623 1 LOG C-0x56245456b010: somedb/someuser@x.x.x.x:54256 closing because: client close request (age=0)
```
Notice the first query takes ~405ms - which includes the call to SET client_encoding, while the 2nd query takes ~202ms as expected.

Log with `Session`:
```
2019-08-12 06:37:08.925 1 LOG C-0x56250c10b010: somedb/someuser@x.x.x.x:54246 closing because: client close request (age=0)
2019-08-12 06:37:08.925 1 DEBUG S-0x56250c15c350: somedb/someuser@y.y.y.y:5432 resetting: RESET ALL;
2019-08-12 06:37:08.929 1 DEBUG C-0x56250c10b010: (nodb)/(nouser)@x.x.x.x:54248 got var: user=someuser
2019-08-12 06:37:08.929 1 DEBUG C-0x56250c10b010: (nodb)/(nouser)@x.x.x.x:54248 got var: database=somedb
2019-08-12 06:37:08.929 1 DEBUG C-0x56250c10b010: (nodb)/(nouser)@x.x.x.x:54248 got var: client_encoding='utf-8'
2019-08-12 06:37:08.929 1 LOG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 login attempt: db=somedb user=someuser tls=no
2019-08-12 06:37:08.929 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 P: got connection: x.x.x.x:54248 -> x.x.x.x:5432
2019-08-12 06:37:08.930 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 logged in
2019-08-12 06:37:08.931 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 pause_client
2019-08-12 06:37:09.134 1 DEBUG S-0x56250c15c350: somedb/someuser@y.y.y.y:5432 S: param: client_encoding = UTF8
2019-08-12 06:37:09.134 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 activate_client
2019-08-12 06:37:09.134 1 DEBUG S-0x56250c15c350: somedb/someuser@y.y.y.y:5432 varcache_apply: SET client_encoding='''utf-8''';
2019-08-12 06:37:09.343 1 DEBUG S-0x56250c15c350: somedb/someuser@y.y.y.y:5432 S: param: client_encoding = UTF8
2019-08-12 06:37:09.343 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 setting client var: client_encoding='UTF8'
2019-08-12 06:37:09.554 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 query time: 622273 us
2019-08-12 06:37:09.554 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 transaction time: 622273 us
2019-08-12 06:37:09.765 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 query time: 209244 us
2019-08-12 06:37:09.766 1 DEBUG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 transaction time: 209244 us
2019-08-12 06:37:09.768 1 LOG C-0x56250c10b010: somedb/someuser@x.x.x.x:54248 closing because: client close request (age=0)
2019-08-12 06:37:09.768 1 DEBUG S-0x56250c15c350: somedb/someuser@y.y.y.y:5432 resetting: RESET ALL;
2019-08-12 06:37:09.977 1 DEBUG S-0x56250c15c350: somedb/someuser@y.y.y.y:5432 S: param: client_encoding = UTF8
```
Notice the first query now takes ~622ms - which is about ~200ms more than with the transaction, while the 2nd takes ~209ms as expected. Inspecting the log I only see `pause_client --> param: client_encoding = UTF8 --> activate_client` as extra vs the `Transaction` mode. 

Bottom line - why is setting client_encoding called twice on `Session`?

On a side note: why is the 1st query reporting the time including all the setup, and not the actual time it took to run (~200ms)? I would expect the setup time to be reported separately and the query time to reflect how much the actual query took without side-effects.

**Comment by petere** [comment]
Apart from the timings, I don't see anything that suggests that client_encoding is set twice. Have you checked the PostgreSQL server log to see if the query actually ends up on the server twice?

**Comment by pyrocks** [comment]
Hi, thanks for replying.
The fact that the log with transaction has a single line `S: param: client_encoding = UTF8` whereas with `session` the same line appears twice with different timestamps seems to suggest it is called twice, no? Or I'm missing something?

**Comment by pyrocks** [comment]
@petere You are right. I set `log_statement` to `all` and redid my test.
Transaction:
```2019-09-25 10:43:38 UTC:x.x.x.x(60074):someuser@somedb:[21112]:LOG: statement: SET client_encoding='''utf-8''';
2019-09-25 10:43:38 UTC:x.x.x.x(60074):someuser@somedb:[21112]:LOG: statement: select 1
2019-09-25 10:43:38 UTC:x.x.x.x(60074):someuser@somedb:[21112]:LOG: statement: select 2
2019-09-25 10:43:39 UTC:x.x.x.x(60074):someuser@somedb:[21112]:LOG: statement: SET client_encoding='''utf-8''';
2019-09-25 10:43:39 UTC:x.x.x.x(60074):someuser@somedb:[21112]:LOG: statement: select 1
2019-09-25 10:43:39 UTC:x.x.x.x(60074):someuser@somedb:[21112]:LOG: statement: select 2
...
```
Session:
```2019-09-25 10:44:51 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: SET client_encoding='''utf-8''';
2019-09-25 10:44:51 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: select 1
2019-09-25 10:44:51 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: select 2
2019-09-25 10:44:52 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: RESET ALL;
2019-09-25 10:44:52 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: SET client_encoding='''utf-8''';
2019-09-25 10:44:52 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: select 1
2019-09-25 10:44:52 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: select 2
2019-09-25 10:44:52 UTC:x.x.x.x(37071):someuser@somedb:[22060]:LOG: statement: RESET ALL;
...
```
So it's just the calls to RESET ALL which I have set as my `server_reset_query`.
So the extra 200ms is the call to RESET ALL before setting the `client_encoding` - but I still believe the times should be reported differently: setup time + actual query time. otherwise it's confusing.
